bootstrap/018_warehouse_updates.sql

```diff
@@ -82,9 +80,22 @@ BEGIN
             let where_clause_complete varchar := (select 'not incomplete and session_end <> to_timestamp_ltz(\'' || :newest_completed || '\')');
             let new_closed number;
             call internal.generate_insert_statement('INTERNAL_REPORTING_MV', 'CLUSTER_AND_WAREHOUSE_SESSIONS_COMPLETE_AND_DAILY', 'INTERNAL', 'RAW_WH_EVT', :where_clause_complete) into :new_closed;
-            insert into INTERNAL.TASK_WAREHOUSE_EVENTS SELECT :run_id, true, :input, OBJECT_CONSTRUCT('oldest_running', :oldest_running, 'newest_completed', :newest_completed, 'attempted_migrate', :migrate, 'migrate', :migrate1, 'migrate_INCOMPLETE', :migrate2, 'new_records', :new_records, 'new_INCOMPLETE', :new_INCOMPLETE, 'new_closed', coalesce(:new_closed, 0))::VARIANT;
+            -- Find oldest session, after the new records have been inserted.
+            let range_min timestamp := (select min(session_end) from (
+                select min(session_end) as session_end from INTERNAL_REPORTING_MV.CLUSTER_AND_WAREHOUSE_SESSIONS_COMPLETE_AND_DAILY
+                union all
+                select min(session_end) as session_end from INTERNAL_REPORTING_MV.CLUSTER_AND_WAREHOUSE_SESSIONS_COMPLETE_AND_DAILY_INCOMPLETE)
+            );
+            output := OBJECT_CONSTRUCT('oldest_running', :oldest_running, 'newest_completed', :newest_completed, 'attempted_migrate', :migrate, 'migrate', :migrate1, 'migrate_INCOMPLETE', :migrate2,
+                'new_records', :new_records, 'new_INCOMPLETE', :new_INCOMPLETE, 'new_closed', coalesce(:new_closed, 0), 'range_min', :range_min, 'range_max', :newest_completed);
         ELSE
-            insert into INTERNAL.TASK_WAREHOUSE_EVENTS SELECT :dt, true, :input, OBJECT_CONSTRUCT('oldest_running', :oldest_running, 'newest_completed', :newest_completed, 'attempted_migrate', :migrate, 'migrate', :migrate1, 'migrate_INCOMPLETE', :migrate2, 'new_records', 0, 'new_INCOMPLETE', 0, 'new_closed', 0)::VARIANT;
+            let range_min timestamp := (select min(session_end) from (
+                select min(session_end) as session_end from INTERNAL_REPORTING_MV.CLUSTER_AND_WAREHOUSE_SESSIONS_COMPLETE_AND_DAILY
```

Comment:
why hand constructing the union as opposed to just using existing object?

Reply:
The higher-level views which union the complete/incomplete tables also include a filter on `session_type = 'CL'` or `session_type = 'WH'`.

I reckon it would be more correct to deduce the exact range of data available for both cluster and warehouse sessions separately. But, that makes the warehouse materialization case different from the rest. Feasible, but more complex.

Reply:
Remember that we should be having a row in the user-facing table for each view that we maintain. The user-facing data should be focused on internal tables/views. So in the case of clsuter session and warehouse session, there should be at least two rows (i think more since there is daily and hourly versions I believe)

Reply:
> there should be at least two rows (i think more since there is daily and hourly versions I believe)

Yeah, I thought I could get away with a single task_log row for one execution of the warehouse_events_maintenance task, but the distinction between cluster_sessions and warehouse_sessions means I have to change this logic.

The daily and hourly variants are easily included when we're actually creating the materialization status view since they are (usually) backed by the same internal table. Clusters/sessions are the exception. I'll break this out.